test_conformance/api/test_api_consistency.cpp

```diff
@@ -0,0 +1,478 @@
+                                    CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT,
+                                    sizeof(b), &b, NULL);
+            test_error(
+                error,
+                "Unable to query CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT");
+            if (b != CL_TRUE)
+            {
+                log_error("DEVICE_QUEUE_SUPPORTED is set but "
+                          "CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT returned "
+                          "CL_FALSE\n");
+                return TEST_FAIL;
+            }
+        }
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_pipes(cl_device_id deviceID, cl_context context,
+                           cl_command_queue queue, int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_PIPE_SUPPORT
+    // May return CL_FALSE, indicating that device does not support Pipes.
+    int error;
+
+    cl_bool pipeSupport = CL_FALSE;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_PIPE_SUPPORT,
+                            sizeof(pipeSupport), &pipeSupport, NULL);
+    test_error(error, "Unable to query CL_DEVICE_PIPE_SUPPORT");
+
+    if (pipeSupport == CL_FALSE)
+    {
+        // clGetDeviceInfo, passing
+        // CL_DEVICE_MAX_PIPE_ARGS,
+        // CL_DEVICE_PIPE_MAX_ACTIVE_RESERVATIONS, or
+        // CL_DEVICE_PIPE_MAX_PACKET_SIZE
+        // Returns 0 if device does not support Pipes.
+
+        cl_uint u = 0;
+
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_MAX_PIPE_ARGS, sizeof(u),
+                                &u, NULL);
+        test_error(error, "Unable to query CL_DEVICE_MAX_PIPE_ARGS");
+        if (u != 0)
+        {
+            log_error("CL_DEVICE_PIPE_SUPPORT returned CL_FALSE, but "
+                      "CL_DEVICE_MAX_PIPE_ARGS returned a non-zero value\n");
+            return TEST_FAIL;
+        }
+
+        error =
+            clGetDeviceInfo(deviceID, CL_DEVICE_PIPE_MAX_ACTIVE_RESERVATIONS,
+                            sizeof(u), &u, NULL);
+        test_error(error,
+                   "Unable to query CL_DEVICE_PIPE_MAX_ACTIVE_RESERVATIONS");
+        if (u != 0)
+        {
+            log_error("CL_DEVICE_PIPE_SUPPORT returned CL_FALSE, but "
+                      "CL_DEVICE_PIPE_MAX_ACTIVE_RESERVATIONS returned a "
+                      "non-zero value\n");
+            return TEST_FAIL;
+        }
+
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_PIPE_MAX_PACKET_SIZE,
+                                sizeof(u), &u, NULL);
+        test_error(error, "Unable to query CL_DEVICE_PIPE_MAX_PACKET_SIZE");
+        if (u != 0)
+        {
+            log_error(
+                "CL_DEVICE_PIPE_SUPPORT returned CL_FALSE, but "
+                "CL_DEVICE_PIPE_MAX_PACKET_SIZE returned a non-zero value\n");
+            return TEST_FAIL;
+        }
+
+        // clCreatePipe
+        // Returns CL_INVALID_OPERATION if no devices in context support Pipes.
+        clMemWrapper mem = clCreatePipe(context, 0, 0, 0, NULL, &error);
```

Comment:
pie_packet_size/pipe_max_packets cannot be 0 while creating pipe to get CL_INVALID_OPERATION

Reply:
Good point - I don't think we've ever specified which error gets returned if two error conditions occur.

I've switched this call to use pipe parameters that would be valid if pipes were supported:

```c
clCreatePipe(context, CL_MEM_HOST_NO_ACCESS, 4, 4, NULL, &error);
```

Note that two error conditions still occur if pipes are not supported, and specifically this call could theoretically return `CL_INVALID_OPERATION` (since no devices in context support Pipes) or `CL_INVALID_PIPE_SIZE` (since the pipe_packet_size exceeds `CL_​DEVICE_​PIPE_​MAX_​PACKET_​SIZE`, which must be zero), but I still think we should keep this check as-is to encourage more uniform behavior.